Integrated  Random Words API

MunawarAliArbisoft opened the pull request:
This pull request introduces several enhancements to the codebase:

1. Integrated Rapid API: which get's random words from the dictionary api along with it's definition and pronunciation and prompt user unique word each time in typing test.

2. Added typings: Added type hints throughout the codebase to improve code clarity and enable static type checking.

3. Implemented .env file: Stored URL and headers in a .env file for better configuration management. Accessing them via the python-dotenv library.

4. Divided programs into modules: Refactored the codebase by grouping related functions into separate modules.

5. Made small optimization to make code clean, optimized and readable 

Please review the changes and provide feedback. These updates aim to improve the overall quality, maintainability, and efficiency of the codebase.

MunawarAliArbisoft committed:
Integrated  Random Words Api

Itezaz-ul-Hassan commented on typing_tester.py:
```diff
@@ -0,0 +1,173 @@
+    url = "https://random-words-api-umber.vercel.app/word"
+
+    headers = {
+        "X-RapidAPI-Key": "f934e406b8msh16f9dee78137dfcp1fab91jsn560d9925327d",
```
Figure out whats the `.env` file, why do we use it and how can we use it in python. Make use of .env here

Itezaz-ul-Hassan commented on typing_tester.py:
```diff
@@ -0,0 +1,173 @@
+        "X-RapidAPI-Host": "random-words-with-pronunciation.p.rapidapi.com"
+    }
+
+    response = requests.get(url, headers=headers)
```
Write complete docstring for this function since its calling external API. Make sure to add the sample response of the API call as well

Itezaz-ul-Hassan commented on typing_tester.py:
```diff
@@ -0,0 +1,173 @@
+      """Displays result of typing tester"""
+    
+      # Gets word, definition and pronunciation from word object
+      word = word_obj['word']
```
Figure out what's the `.get` method of the dictionary does and why do we use it instead of using brackets

MunawarAliArbisoft replied:
Resolved

Itezaz-ul-Hassan commented on typing_tester.py:
```diff
@@ -0,0 +1,173 @@
+      stdscr.refresh()
+      key = stdscr.getch()
+
+    #   if ord(key) != 'q':
```
Whenever we write code, we shouldn't add commented code if it's no longer useful.

Itezaz-ul-Hassan reviewed (changes requested):
Make use of typings here as well

MunawarAliArbisoft committed:
Add typings, .env file, divide programs into modules, and make small optimization/clean code changes

In this commit I added typings, created .env file to store environment variables for api, divided programs into modules to make codebase organized and clean and made various small optimizations and improvements to enhance performance and readability.

Itezaz-ul-Hassan commented on .env:
```diff
@@ -0,0 +1,3 @@
+API_URL = "https://random-words-api-umber.vercel.app/word"
```
Never ever make this mistake again.

DO NOT COMMIT OR PUSH `.env` TO THE CODE. Add it into .gitignore

MunawarAliArbisoft replied:
ok

Itezaz-ul-Hassan commented on .env:
```diff
@@ -0,0 +1,3 @@
+API_URL = "https://random-words-api-umber.vercel.app/word"
```
Make a .env.sample or .env.dist, figure out what these are for and why it's good practice to have them in your code

Itezaz-ul-Hassan commented on calculate_score.py:
```diff
@@ -0,0 +1,30 @@
+
```
Always format python code using a linter, be it `pylint`, `flake8` or `black`

Itezaz-ul-Hassan commented on calculate_score.py:
```diff
@@ -0,0 +1,30 @@
+
+def cal_accuracy_score(given_word: str, 
+                       user_text: str, no_of_backspaces: int) -> float:
```
See rules for formatting

Itezaz-ul-Hassan replied:
This function returns a Tuple. Figure out what are tuples. Also change the return type of this function to Tuple and then add relevant types in it

Itezaz-ul-Hassan commented on screens.py:
```diff
@@ -0,0 +1,112 @@
+
+        try:
+            if ord(key) == DELETE_KEY_ASCII:
+                if len(user_text_list) > 0:
```
```suggestion
                if len(user_text_list):
```

Itezaz-ul-Hassan commented on screens.py:
```diff
@@ -0,0 +1,112 @@
+            if ord(key) == DELETE_KEY_ASCII:
+                if len(user_text_list) > 0:
+                        user_text_list.pop()
+                        no_of_backspaces = no_of_backspaces + 1
```
```suggestion
                        no_of_backspaces += 1
```

MunawarAliArbisoft replied:
Resolved

Itezaz-ul-Hassan commented on screens.py:
```diff
@@ -0,0 +1,112 @@
+    stdscr.getkey()
+    typing_tester_screen(stdscr) # Test Screen where user input is evalauted
+
+def typing_tester_screen(stdscr):
```
What does this function return?

MunawarAliArbisoft replied:
It's a void function that displays the screen where user performs typing test

Itezaz-ul-Hassan commented on screens.py:
```diff
@@ -0,0 +1,112 @@
+
+            elif len(user_text_list) < len(word_obj.get('word')):
+                user_text_list.append(str(key))
+        except:
```
Add specific exception type

Itezaz-ul-Hassan reviewed (changes requested):
Always be persistent and use only one kind of quotes throughout the entire codebase either single quotes or double. Don't mix them up.

MunawarAliArbisoft committed:
Add .env to .gitignore, created .env.dist and format the code

Created .env.dist with sample env variables and add .env to .gitignore, formatted code to make look clean and readable, change return type of cal_accuracy_score to tuple.

MunawarAliArbisoft committed:
Merge branch 'main' into typing-tester-api-integration

Itezaz-ul-Hassan commented on screens.py:
```diff
@@ -0,0 +1,112 @@
+from calculate_score import cal_accuracy_score, cal_time_score
+
+
+def start_screen(stdscr):
```
```suggestion
def start_screen(stdscr) -> None:
```

MunawarAliArbisoft committed:
fix: Add Return Type to start_screen func

Co-authored-by: Itezaz ul Hassan <109591867+Itezaz-ul-Hassan@users.noreply.github.com>
